### ISSUE TITLE
Error when closing a menu by clicking on any event in the log section with date or timestamp activated

### ISSUE THREAD

**Issue by DrUm78** [EventGhost/EventGhost#348]
An error occured when closing a menu by clicking on any event in the log part with date or timestamp activated

1. Open EventGhost v0.5.0-rc4, v0.5.0-rc5 or  v0.5.0-rc6 (not happening on stable 0.4.1.r1722 BTW)
2. Activate Datestamp or Timestamp in 'View' menu (does not occur if none of them are enabled)
3. Open any menu ('View' for instance)
4. Click in the log section on any event to close the menu
5. Notice that you get an error

Please see the attached screenshot.

![timestamp_error](https://user-images.githubusercontent.com/46139740/50427350-b45b0200-08a5-11e9-92e6-c8e84a150d43.jpg)

**Comment by kdschlosser** [comment]
now if you do not have the date time stamp activated this does not happen?

**Comment by Drumaster78** [comment]
Yes, it only happens with date or timestamp enabled (one of them or both of them), otherwise everything is fine. :)

**Comment by topic2k** [comment]
I can't reproduce the error (with RC6 on win10 and win7), so i guess it's something with your installation.
Do you have any special system tools running?
Is it a slow computer?
Was it a clean install?
Any other python installations?

The only idea i have ATM is to uninstall and completly remove all EG files (**make a backup**) from `%ProgramFiles%\EventGhost`, `%ProgramData%\EventGhost` and `%APPDATA%\EventGhost`. Then install EG and try again (only changing the stamp settings before changing anything else).

**Comment by Drumaster78** [comment]
I just uploaded a video (unzip it first)= to show you the exact steps :
[stamp_error.zip](https://github.com/EventGhost/EventGhost/files/2709949/stamp_error.zip)

**Comment by topic2k** [comment]
Yes, i did the exact steps as you mentioned. Please do me a favour and open `%ProgramFiles%\EventGhost\eg\Classes\MainFrame\LogCtrl.py` in notepadd++ and change lines 254-258 from
```python
        return (
            strftime(self._datetime_fmt, localtime(when)) +
            indent * self.indent +
            line
        )
```
to
```python
        try:
            return (
                strftime(self._datetime_fmt, localtime(when)) +
                indent * self.indent +
                line
            )
        except:
            eg.PrintTraceback()
            print repr(line)
            print repr(when)
            print repr(indent)
            print repr(self.indent)
```
I just wanna see if we can catch the error and get some more information from it.

**Comment by Drumaster78** [comment]
Ok I just tried and the error seems to be the same. What do you want me to provide? Here is the screenshot of what I get after modifying LogCtrl.py and launching in debug mode (option Debug in View menu does the same thing right?).
![sans titre](https://user-images.githubusercontent.com/46139740/50443875-90c7a400-0905-11e9-8d60-9d0902c692fb.jpg)


BTW, I forgot to mention that I am on Windows 10 Professional build 1809.

**Comment by kdschlosser** [comment]
I am running on Windows 7 and I cannot reproduce this either.

We have no doubt it is happening. we simply need to discover the cause of the problem.

**Comment by kdschlosser** [comment]
OK I watched your video. 

In order for us to compare apples to apples. we need you to exactly follow all instruction step by step.
Because if you do not we will not be able to isolate the issue. We need to have a baseline in order to know where abouts the error may be coming from.

I noticed that in your video that was not what EG should look like after a clean fresh installation. just to name a few things the window was the wrong size,  the log had been cleared.

One of the things we have learned doing this. It is always the smallest and simplest thing that gets overlooked is what causes the most complex problems. 

It could be something else that you have clicked on that is the actual cause of the issue and not where we are currently looking. We want to get this problem solved but we need you to be very diligent in repeating the exact steps we outline with no deviation.

**Comment by Drumaster78** [comment]
I totally know that as I've been working myself as a QA in the video games industry for almost 15 years, don't worry. I repeated again what you said and removed all files from %ProgramFiles%\EventGhost, %ProgramData%\EventGhost and %APPDATA%\EventGhost, launched the software WITHOUT resizing anything or cleaning the log. I only set Datestamp and the issue immediately appeared again.

Please tell me what to do to provide you with more useful information, I'll do it without any problem, thanks!

**Comment by kdschlosser** [comment]
now here comes a possible reason as to why the issue is occurring.
which is something that we did not know about because of changes you made so that we would be able to read what was going on.

You have a different language this time around. which is going to make me think it could be locale related.

setting the language in EG works independently from the Windows user language and locale.
any chance of getting you to change the locale in windows? I would like to know if this issue goes away if you do.

Control Panel -> Region and Language

Formats Tab
change the Format to English (United States)

Location tab
change the current location to United States

Administrative tab
click on the Change system locale button
a dialog will open.
change the Current system locale to English (United States)

You will probably have to restart your computer

**Comment by kdschlosser** [comment]
another note. we have changed EG to be more locale friendly. and when doing this we also uncovered some bugs in a piece of software that was not written by us. It is called wxPython (wxWidgets) we have been correcting the issues as they come up. Actually I am right in the middle of rewriting the whole language system in EG it will support over 175 languages and some 150 countries. and it will also build the translations for plugins as well as the core of EG on the fly. 

This is a pretty large undertaking and it will be a while before it gets included into EG.

**Comment by topic2k** [comment]
Not sure if we are here facing the same problem as in wxWidgets/Phoenix#992 (which links to [wxWidgets#18221](https://trac.wxwidgets.org/ticket/18221#ticket)).

**Comment by kdschlosser** [comment]
I do not know what is so special about you setup that is causing this issue. The menu closing when the focus shifts to something else other then the Menu is controlled internally by wxPython. Maybe I can monkey patch the thing. So at the very least I can absorb the error. It appears as tho the menu still does close properly. there is something off timing wise.

Let me see what I can do. It is super difficult to find a solution when we are unable to replicate the problem. But if you are willing to spend the time to find the solution I will be more then willing to throw some test code at ya. It may take a while for me to nail down the exact mechanics inside of wxPython that handle the menu closing. 

I only say we will have to fix this ourselves because the folks over there at wxPython and wxWidgets seem to not pay attention to bug reports. In the last post where i said w are correcting the issues. well these issues have been reported to wxPython and wxWidgets. 10 years ago. and nothing has been done to correct them. So we simply will not rely on them to fix anything. We have to find a way to make it work correctly. 

I think that @topic2k and I have done a smashing job thus far finding ways to work around issues in other software. I will say this. I never giv up on trying to isolate a problem and repair it. It may take a while sometimes. But so far I have not said to anyone. "Sorry that's just the way it is so you have to deal with it" I have always found a solution.

**Comment by Drumaster78** [comment]
Thank you for your answer. I really appreciate your dedication about fixing bugs, and EG is a great piece of software, so damn useful for a lot of things. :)

About the issue itself, it appears on 3 different computers (2 of mine and one at work), all with EG RC6 on Windows 10 Pro build 1809, so I don't know how they could be some 'special setup'. BTW I will try tomorrow on a Windows 7 Ultimate to see how it behaves.

Anyway, I'm available for beta testing your test code, no problem about that, just send it to me. However, this issue had some workarounds and is not a blocking one (it does not make the app crash or freeze).

**Comment by kdschlosser** [comment]
no i know that the program is still functional. but it is still technically speaking "broken" 1 error is 1 error to many.

**Comment by Drumaster78** [comment]
Ok, after testing on my old laptop with core i5 and Win 7 Ultimate SP1 in French language/region, the bug does not occur, so that's interesting. This system has not been updated with Windows Update for years FYI. 

Don't hesitate if you want more details.

**Comment by kdschlosser** [comment]
I am almost 100% sure this problem is being caused by something internal to wxPython. I am trying to isolate the exact component and see if I can "monkey patch" the problem. trying to locate specific pieces of the source for wxPython is not an easy task. because it is actually broken into 2 extremely large projects wxPython and wxWidgets for  first being an extension of the last. I need to locate exactly what I have to override and change the behavior of. So do not expect anything from me for a few days at the very least.

@topic2k if you have any input on this please make some noise.

**Comment by topic2k** [comment]
@kdschlosser did you see my [post](https://github.com/EventGhost/EventGhost/issues/348#issuecomment-450217305) 2 days ago? It's probably an issue in wxWidgets, but i don't understand why it doesn't happen on my Win10pro (1809). Maybe because i'm in a VirtualBox?

**Comment by kdschlosser** [comment]
I did see what you had posted. Tho i do not know if the 2 are linked in some manner. as the error in that bug report is completely different. It also is not going to help us if and when they fix it because they are no longer supporting Python 2x variants of wxPython so the only version that is going to see any fixes is going to be wxPython 4.x

So we still need to locate a wy to handle this problem ourselves. The biggest problem is we do not know the exact cause of the problem. This is what is really burning my biscuit. without the piece of information is going to be a real hard bugger to fix a problem when we don't even know exactly what the problem is. 

Now I am not sure as to why this has all of a sudden come up as well. Windows 10 is not exactly new. So it has to be something setting related to @Drumaster78 's specific setup. What exactly is different about @Drumaster78 's specific Windows setup is the question that we need to find out.

I cannot think of anything "Setting" wise in Windows that can cause this kind of a problem. Maybe it is another Application causing the problem. Possibly one that monitors the Windows Desktop and GUI ??? 

What has baffled me is the fact that the error states that it is trying to close a menu that is already closed. Because the open menu from the menu bar is closed successfully this means that wxPython did close it correctly. So if the error is about that menu. then why is wxPython trying to close it twice? and if it is not trying to close it 2 times the question becomes which menu does wxPython think is open? we are assuming the error is from the drop down menu. but maybe it is not. it could be for some other menu. 

What I think we need to do is patch wxPython to spit out the menu ID's when a menu gets opened. and closed. this way we can see which one is getting "stuck" and then go from there.

Virtual Box should not cause any change in behavior like this.

**Comment by Joyomojo** [comment]
I am using RC5 and got the same problem.  Win10 1803 German
One thing that i noticed is that the error behaves a bit different depending on where you click and what is marked.
For example:
-no row is marked > open menu > close menu by clicking the empty space below the log  = no error
-no row is marked > open menu > close menu by clicking on a row =  error
-a row is marked > open menu > close menu by clicking on a row =  error
-a row is marked > open menu > close menu by clicking the empty space below the log =  error (same message) and the text of the row that was marked disappears, icon and background color stays.
Text reappears clicking any of the rows but not when i click the empty space below.

Probably not related but is it intented that the right click menu takes a double click to open when i click below the log instead of a single click when i click one of the log rows?

**Comment by kdschlosser** [comment]
what menu opens when you right click outside of the log window?
can you take a screen shot of that please.. alt + printscreen and then paste it directly into a github comment.

also is it the same error that appears each time you encounter it when the menu closes??

also is it only the dropdown menus from the top that cause this error to appear. or does it also happen when you right click in the log (context menu) and then click elsewhere to close the context menu

**Comment by Joyomojo** [comment]
![image](https://user-images.githubusercontent.com/2333716/50573247-b60a5580-0dd0-11e9-818d-d872a4e76f81.png)
It is the context menu. Just seems weird that it needs a double right click to open there.

Same error message everytime and only the dropdown menus from the top seem to cause it.

**Comment by topic2k** [comment]
@Drumaster78 can please try [this build](https://ci.appveyor.com/api/buildjobs/e333p0ajap1l9a4f/artifacts/_build%2Foutput%2FEventGhost_WIP-2019.01.06-12.43.25_Setup.exe). It changes where the menu event bindings belong to. This fixes at least the context menu bug @Joyomojo described above.

**Comment by Drumaster78** [comment]
Still the same issue for me with your test build, even aftere a clean install. :(

**Comment by topic2k** [comment]
Ok, next round :smiley:  [This build](https://ci.appveyor.com/api/buildjobs/lg62j6y9qaoa7do4/artifacts/_build%2Foutput%2FEventGhost_WIP-2019.01.06-13.32.11_Setup.exe) uses a newer version of the GUI system (wxPython 4.0.3). There can still be other errors, but at least the standard install works for me. I have to review the plugins. It's still WIP.

**Comment by Drumaster78** [comment]
After a few more testing, I just noticed that adding plugins seems to be broken from rc5 version. I realized this only now because I added all my plugins previously in rc4.

Steps:
- Install rc5 or rc6
- Try to add 'Keyboard' plugin (in 'Input Devices' section) for instance
- Notice that the 'OK' button is grayed in the plugin window
- Just uncheck the option 'Universal modifiers' in the window
- Notice that you can now add the plugin

**Comment by topic2k** [comment]
The last build which i linked above is a WIP forwards to wxPython 4. It's not up to date with all fixes we did in RC5/6. That's why see the bugs again.

**Comment by kdschlosser** [comment]
I thought wxPython 4 was only for Python 3+

**Comment by kdschlosser** [comment]
and there is a HUGE absolutely HUGE amount of differences between wxPython 3 and wxPython 4.

remember I started to make that wrapper to fix all of the issues. I have a lot of them handled with that thing. but not all.

**Comment by topic2k** [comment]
No, wx4 is also for py2. Yes, there are lot changes and i already adopted nearly all code we have in our repo. It's been a while since i've done that and i can't remember if i updated all plugins. I will continue work on this in the next days so we can eventually do a 0.6 pre-release :D

**Comment by kdschlosser** [comment]
Oh cool. then have a look at the last PR i did. It is going to take a but for eventghost to load for the first time. I have been trying to get a dialog to display. But I have had no such success as of yet. so be patient when you first run it.

**Comment by kdschlosser** [comment]
the other thing is. you do know it is only you and I to update the 400+ plugins EG has.. (sigh)
